### PR TITLE
Build cache

### PR DESCRIPTION
**Pull request by Crell**
Still a few things here to clarify with the language experts.

### CODE REVIEW

File: src/configuration/app/build.md [hunk 1]
```diff
@@ -23,6 +25,113 @@ build:
...
+            allow_stale: true
+{%- language name="Node/NPM", type="yaml" -%}
+build:
+    caches:
+        node_modules:
+            watch: package-lock.json
+{%- language name="Python", type="yaml" -%}
+build:
+    caches:
+        what_dir_is_this:
```
[the remark this answers is not in the document]
Comment (demosdemon) [reply src/configuration/app/build.md]: this should probably be
```
build:
    caches:
        .global/lib/python3.7/site-packages:
            watch:
                - Pipfile
                - Pipfile.lock
```

File: src/configuration/app/build.md [hunk 2]
```diff
@@ -23,6 +25,113 @@ build:
...
+            watch:
+                - Pipfile
+                - Pipfile.lock
+{%- language name="Ruby", type="yaml" -%}
+build:
+    caches:
+        what_dir_is_this:
```
[the remark this answers is not in the document]
Comment (demosdemon) [reply src/configuration/app/build.md]: this should probably be whatever this returns
```sh
$ platform ssh -p <project> -e <environment> 'gem environment | grep -F "SPEC CACHE DIRECTORY"'
  - SPEC CACHE DIRECTORY: /app/.gem/specs
```

File: src/configuration/app/build.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (pjcdawkins) [reply src/configuration/app/build.md]: Keys shouldn't start with `.`, and that key can just be the name of the cache (it's also the directory by default, but doesn't have to be). I suggest:

```yaml
build:
    caches:
        pip:
            dir: '.global/lib/python3.7/site-packages'
            watch:
                - Pipfile
                - Pipfile.lock
```

[the remark this answers is not in the document]
Comment (Crell) [reply src/configuration/app/build.md]: Thanks, I've added this.

File: src/configuration/app/build.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (Crell) [reply src/configuration/app/build.md]: Got it, thanks.

**Comment by lsmith77** [comment]
guess it would make sense to mention that changes to the cache config (ie. removing a dir) will not lead to cleaning up that dir automatically. more importantly it seems dirs will be restored if they were once cached. not sure if this could cause issues between environments.

at any rate it would make sense to mention `platform project:clear-build-cache`

**Comment by jmdrawneek** [comment]
@Crell is this PR dead? I would love to see this feature on Platform.

**Comment by matthiaz** [comment]
@jmdrawneek Definitely not dead. It's delayed a bit, but it is coming.

File: src/configuration/app/build.md [hunk 3]
```diff
@@ -4,7 +4,9 @@ The `.platform.app.yaml` file provides a number of ways to control how an applic
 
 ## Build
 
-The `build` defines what happens when building the application.  Its only property is `flavor`, which specifies a default set of build tasks to run. Flavors are language-specific.
+The `build` defines what happens when building the application.  It has two properties, `flavor`, which specifies a default set of build tasks to run, and `caches`, which allows fine-grained control of the build cache.
+
+The available flavors are language specific:
```
Comment (RazzyEatsRelish) [inline src/configuration/app/build.md:9]:
```
```
```
`~~~~`
```
```
```

File: src/configuration/app/build.md [hunk 4]
```diff
@@ -4,7 +4,9 @@ The `.platform.app.yaml` file provides a number of ways to control how an applic
 
 ## Build
 
-The `build` defines what happens when building the application.  Its only property is `flavor`, which specifies a default set of build tasks to run. Flavors are language-specific.
```
Comment (RazzyEatsRelish) [inline src/configuration/app/build.md:7]: # @JGrubb @dubcanada ~~~~```

```~~~~@matthiaz @jmdrawneek @lsmith77 @Crell @mrkschan @hitfactory @dubcanada @achton @daharon @spk @robertDouglass

**Review by RazzyEatsRelish — APPROVED** [review]

**Comment by CollierCZ** [comment]
Closing for now. Will reopen or create a new PR when the feature is closer to done.

**Comment by CollierCZ** [comment]
Closing for now. Will reopen or recreate when the feature is closer to done.

**Comment by CollierCZ** [comment]
Will redo when the feature stops being on hold.

**Pull request closed without merging**